Add remaining services to `cartesi-rollups-node`

torives opened the pull request:
Closes #139

torives committed:
refactor: replace Service with its implementation

Currently, there isn't a need to have multiple implementations of the Service interface. Therefore, this commits replaces it with its only implementation to keep things simple.

torives committed:
refactor: create background context in main

# Conflicts:
#	internal/services/service.go

torives committed:
fix: correctly assess process exit reason

We were incorrectly comparing the exit code with the SIGTERM signal. With this fixed, service.Start() won't return an error if the process was killed because its context was canceled

gligneul commented on internal/services/service.go:
```diff
@@ -45,33 +46,57 @@ func (s simpleService) Start(ctx context.Context) error {
 	// start services
-	ctx, cancel := context.WithCancel(context.Background())
-	exit := make(chan struct{})
+	startedServicesCount := 0
```
We can replace part of this logic with a WaitGroup.
https://pkg.go.dev/sync#example-WaitGroup

I will write a PR on top of your proposing something.

gligneul commented on internal/services/service.go:
```diff
@@ -104,17 +140,76 @@ func Run(services []Service) {
+}
+
+func healthcheckPort(serviceName string) string {
+	if serviceName == "state-server" {
```
Please create a separate function for the state server.
```
func stateServerHealthcheckPort() string {
```

gligneul commented on test/fakeservice/main.go:
```diff
@@ -0,0 +1,10 @@
+// This file creates a dummy webserver with the sole pupose of being used
```
Put this file inside of ./internal/services/mockservice/

gligneul commented on internal/services/service_test.go:
```diff
@@ -0,0 +1,144 @@
+	buildFakeService()
+}
+
+func TestService(t *testing.T) {
```
The tests left a zombie fake-service running on my machine. We need to make sure we are cleaning all resources.

gligneul commented on internal/services/service.go:
```diff
@@ -104,17 +140,76 @@ func Run(services []Service) {
-		binaryName:  "cartesi-rollups-graphql-server",
+	AdvanceRunner = Service{
+		name:            "advance-runner",
+		binaryName:      "cartesi-rollups-advance-runner",
+		healthcheckPort: healthcheckPort("advance-runner"),
+	}
+	AuthorityClaimer = Service{
+		name:            "authority-claimer",
+		binaryName:      "cartesi-rollups-authority-claimer",
+		healthcheckPort: healthcheckPort("authority-claimer"),
+	}
+	Dispatcher = Service{
+		name:            "dispatcher",
+		binaryName:      "cartesi-rollups-dispatcher",
+		healthcheckPort: healthcheckPort("dispatcher"),
+	}
+	GraphQLServer Service = Service{
+		name:            "graphql-server",
+		binaryName:      "cartesi-rollups-graphql-server",
+		healthcheckPort: healthcheckPort("graphql-server"),
+	}
+	Indexer Service = Service{
+		name:            "indexer",
+		binaryName:      "cartesi-rollups-indexer",
+		healthcheckPort: healthcheckPort("indexer"),
+	}
+	InspectServer = Service{
+		name:            "inspect-server",
+		binaryName:      "cartesi-rollups-inspect-server",
+		healthcheckPort: healthcheckPort("inspect-server"),
 	}
-	Indexer Service = simpleService{
-		serviceName: "indexer",
-		binaryName:  "cartesi-rollups-indexer",
+	StateServer = Service{
+		name:            "state-server",
+		binaryName:      "cartesi-rollups-state-server",
+		healthcheckPort: healthcheckPort("state-server"),
 	}
 )
```
I feel that this is in the wrong place. This pkg doesn't need to know about the instantiation of the node. Probably, this should be on the cmd or a `node` pkg.

torives replied:
> This pkg doesn't need to know about the instantiation of the node

I agree. I'll put it in `validator.go` for now and after we get rid of Cobra I'll revisit this.

torives replied:
Actually, I changed my mind. There's enough boilerplate code already to justify a `node` package.

gligneul commented on internal/services/service.go:
```diff
@@ -1,38 +1,39 @@
+	DefaultServiceTimeout  = 15 * time.Second
+	DefaultDialInterval    = 100 * time.Millisecond
+	DefaultStateServerPort = "50051"
+	DefaultHealthcheckPort = "8080"
```
I don't like using a default port because it leads to port conflicts.

torives replied:
> Can't we make it explicit every time?

Yes, we can. Also, It's better to keep it all together so it becomes easier to throw away when #155 is merged.

gligneul committed:
test: remove Rust binary dependency

gligneul commented:
See https://github.com/cartesi/rollups-node/pull/162

gligneul commented on internal/services/service_test.go:
```diff
@@ -0,0 +1,144 @@
+		panic(err)
+	}
+
+	cmd := exec.Command("go", "build", "-o", "build/fake-service", "test/fakeservice/main.go")
```
I don't like the idea of putting this binary in the build dir because this mixes up compilation results with committed files.
I see two options:
1. During the test, create a temp dir, use it as an output dir, and remove this dir at the end of the test.
2. Create a new directory in the project called `bin` or `target` that contains the compilation output.

torives committed:
refactor: improve system cancelation

Uses context.Context to cancel services and sync.WaitGroup to wait for them to finish, making the code more idiomatic and easier to read

torives committed:
feat: add remaining Rust services to Go binary

torives committed:
test: rewrite tests with testify

gligneul commented on internal/services/service_test.go:
```diff
@@ -0,0 +1,147 @@
+
+// Builds the fake-service binary and adds it to PATH
+func (s *ServicesTestSuite) buildFakeService() {
+	rootDir, err := filepath.Abs("../../")
```
Why do we need this?

torives replied:
Indeed, `rootDir` is not needed anymore. I'll remove it.

gligneul reviewed (changes requested):
Looks good. Minor comments.

torives commented on internal/node/node.go:
```diff
@@ -0,0 +1,89 @@
+	"github.com/cartesi/rollups-node/internal/services"
+)
+
+var ValidatorServices = []services.Service{
+	*StateServer,
+	*AdvanceRunner,
+	*AuthorityClaimer,
+	*Dispatcher,
+	*GraphQLServer,
+	*Indexer,
+	*InspectServer,
+}
```
Thinking this through, I see no point in using pointers (pun intended). Let's go with values 😉

torives commented on internal/node/node.go:
```diff
@@ -0,0 +1,89 @@
+		}
+		return port
+	} else {
+		return "8080"
```
Good point. I'll remove it.

torives commented on internal/services/service_test.go:
```diff
@@ -0,0 +1,147 @@
+
+func (s *ServicesTestSuite) TearDownSuite() {
+	err := os.RemoveAll(s.tmpDir)
+	if err != nil {
+		panic(err)
+	}
```
I hadn't thought about using assert/require for functions that are not proper tests. After trying it I found that using `panic` provides a better description of what happened since it prints the error message instead of the error object itself. 

I would rather keep this as is, but we can discuss this further if you like.

GMKrieger reviewed (dismissed):
LGTM

gligneul commented on build/docker-compose.yml:
```diff
@@ -38,7 +65,19 @@ services:
       INDEXER_HEALTHCHECK_PORT: 8083
       DAPP_CONTRACT_ADDRESS_FILE: /deployments/localhost/dapp.json
```
This will change soon when we have the unified config that Renan is working on.
